dfs/registry/listener.go

```diff
@@ -156,11 +160,32 @@ func (l *RegistryListener) Spawn(shutdown <-chan interface{}, id string) {
 					if err := l.docker.TagImage(node.Image.UUID, registrypath); err != nil {
 						glog.Warningf("Could not tag %s as %s: %s", node.Image.UUID, registrypath, err)
 						node.PushedAt = time.Unix(0, 0)
-					} else if err := l.docker.PushImage(registrypath); err != nil {
-						glog.Warningf("Could not push %s: %s", registrypath, err)
-						node.PushedAt = time.Unix(0, 0)
 					} else {
-						node.PushedAt = time.Now().UTC()
+						var err error
+						if node.AfterCommit { //Handle the special case where this is the first push after a commit
+							glog.V(1).Infof("This is a post-commit push, so we will Push, Remove, Pull")
+							var newID	string
+							if newID, err = l.docker.PushImageAfterCommit(registrypath); err != nil {
+								glog.Errorf("Could not perform post-commit push for %s: %s", registrypath, err)
+								node.NewID = ""
+								node.AfterCommit = true
+								node.AfterCommitPushFailed = true	
+								node.PushedAt = time.Unix(0, 0)
+							} else {
+								node.NewID = newID
```

Comment:
utils.go - SetRegistryImageAfterCommit watches the node until the push is complete, then returns the new ID to the facade, which re-pushes with the correct ID.